src/container/srv_target.c

```diff
@@ -37,39 +37,47 @@
 
 /* Per VOS container aggregation ULT ***************************************/
 
-/*
- * VOS aggregation should try to avoid aggregating in the epoch range where
- * lots of data records are pending to commit, so the highest aggregate epoch
- * will be:
- *
- * current HLC - (DTX batched commit threshold + buffer period)
- */
-#define DAOS_AGG_THRESHOLD	(DTX_COMMIT_THRESHOLD_AGE + 10) /* seconds */
-
-#define DAOS_AGG_LAZY_RATE	50 /* ms */
+static inline struct sched_request *
+cont2req(struct ds_cont_child *cont, bool vos_agg)
+{
+	return vos_agg ? cont->sc_agg_req : cont->sc_ec_agg_req;
+}
 
-bool
+int
 agg_rate_ctl(void *arg)
 {
 	struct agg_param	*param = arg;
 	struct ds_cont_child	*cont = param->ap_cont;
 	struct ds_pool		*pool = cont->sc_pool->spc_pool;
-	struct sched_request	*req = param->ap_req;
-
-	if (dss_ult_exiting(req))
-		return true;
-
-	switch (pool->sp_reclaim) {
-	case DAOS_RECLAIM_DISABLED:
-		return true;
-	default:
-		if (dss_xstream_is_busy() &&
-		    sched_req_space_check(req) == SCHED_SPACE_PRESS_NONE)
-			sched_req_sleep(req, DAOS_AGG_LAZY_RATE);
-		else
-			sched_req_yield(req);
-		return false;
+	struct sched_request	*req = cont2req(cont, param->ap_vos_agg);
+
+	/* Abort current round of aggregation */
+	if (dss_ult_exiting(req) || pool->sp_reclaim == DAOS_RECLAIM_DISABLED)
+		return -1;
+
+	/* System is idle, let aggregation run in tight mode */
+	if (!dss_xstream_is_busy()) {
+		sched_req_yield(req);
+		return 0;
+	}
+
+	/*
+	 * When it's under space pressure, aggregation will continue run in slack
```

Comment:
Aggregation runs in tight mode when the system is idle,  when system is busy on processing IO, we want to yield more often (slack mode) to minimize the performance impaction.

Reply:
I understood the current policy for tight and slack, but my concern is whether such policy is reasonable or not to handle space pressure. Because if system has space pressure and if we give more CPU cycles to the regular IO processing, does it means the risk of system out of space is more large? My understand is that, under such case, we need to restrict the regular IO and give more CPU cycles to VOS aggregation and GC to release more space efficiently, that is tight mode instead of slack mode.

Reply:
Yes, CPU apportioning is done by scheduler. There is also a pending patch for the CPU apportioning improvement: https://github.com/daos-stack/daos/pull/8513 , let me add you as reviewer.

Reply:
Then that is fine for me.